docs/src/explanations/advanced-data-types.md

```diff
@@ -0,0 +1,141 @@
+---
+layout: docs
+title:  "Advanced Data Types"
+section: "chisel3"
+---
+
+# Advanced Data Types
+
+The Scala compiler cannot distinguish between Chisel's representation of hardware such as `false.B`, `Reg(Bool())`
+and pure Chisel types (e.g. `Bool()`). You can get runtime errors passing a Chisel type when hardware is expected, and vice versa.
+
+## Scala Type vs Chisel Type vs Hardware
+
+The *Scala* type of the Data is recognized by the Scala compiler, such as `Decoupled[UInt]` or `MyBundle`, where 
+```
+MyBundle(w: Int) extends Bundle {val foo: UInt(w.W), val bar: UInt(w.W)}
+```
+
+The *Chisel* type of a Data is all the fields actually present, by names, and their types including widths and directions.
```

Comment:
What I'm wanting to change in the conceptual model is that Chisel types define relative direction (flipped/aligned) of their fields, but Chisel types do NOT define the actual direction of their leafs.

This means a few things.

1. Absolute direction of a leaf field is a property of the hardware it is bound as, NOT as a property of the Chisel type. This is why I want to add `Incoming` and `Outgoing`, because they define the absolute direction of leafs of their Chisel types (subject to relative flips along the path from root-to-leaf field).
2. Along those lines, when Chisel types define absolute direction, `Wire` does not make sense conceptually in this model because a wire's direction is usually defined by how it is being used/connected. In fact, only ports really care about absolute direction.
3. Producer/consumer (who drives who in a connection) is a property of the connection operator, not a property of the "absolute direction" of things, because again that doesn't make sense for wires. *This is a desirable property* because the connection semantics work the same on all hardware of the same type (be it wires or ports or registers etc.)
4. This is why Wes's connection operators are so useful, because they are consistent regardless of the underlying bound hardware, while still being powerful enough to describe many use cases e.g. connecting to monitors.

Reply:
I like these directions, but as it is today for documenting, the ChiselType I believe *does* capture the directionality? Or shoudl I just say `directionality*    *it's complicated`?